[CI Failure]: ImportError: cannot import name 'BuildDependencyInstallError' from 'pip._internal.exceptions'

### CI Test Type

Frontend

### Which CI step failed?

Merge develop and set up dependencies

### Stacktrace

```shell
Check pip installation and fix if needed
1s
Run set -e
  Checking if pip is properly installed...
  Traceback (most recent call last):
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/runpy.py", line 196, in _run_module_as_main
      return _run_code(code, main_globals, None,
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/runpy.py", line 86, in _run_code
      exec(code, run_globals)
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/__main__.py", line 22, in <module>
      from pip._internal.cli.main import main as _main
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/main.py", line 11, in <module>
      from pip._internal.cli.autocompletion import autocomplete
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/autocompletion.py", line 12, in <module>
      from pip._internal.cli.main_parser import create_main_parser
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/main_parser.py", line 9, in <module>
      from pip._internal.build_env import get_runnable_pip
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/build_env/__init__.py", line 8, in <module>
      from pip._internal.build_env.installer import (
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/build_env/installer.py", line 14, in <module>
      from pip._internal.exceptions import (
  ImportError: cannot import name 'BuildDependencyInstallError' from 'pip._internal.exceptions' (/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/exceptions.py)
  pip is broken. Clearing pip cache...
  Traceback (most recent call last):
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/runpy.py", line 196, in _run_module_as_main
      return _run_code(code, main_globals, None,
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/runpy.py", line 86, in _run_code
      exec(code, run_globals)
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/__main__.py", line 22, in <module>
      from pip._internal.cli.main import main as _main
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/main.py", line 11, in <module>
      from pip._internal.cli.autocompletion import autocomplete
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/autocompletion.py", line 12, in <module>
      from pip._internal.cli.main_parser import create_main_parser
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/main_parser.py", line 9, in <module>
      from pip._internal.build_env import get_runnable_pip
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/build_env/__init__.py", line 8, in <module>
      from pip._internal.build_env.installer import (
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/build_env/installer.py", line 14, in <module>
      from pip._internal.exceptions import (
  ImportError: cannot import name 'BuildDependencyInstallError' from 'pip._internal.exceptions' (/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/exceptions.py)
  Reinstalling pip and dev dependencies via install_python_dev_dependencies.py...
  Traceback (most recent call last):
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/runpy.py", line 196, in _run_module_as_main
      return _run_code(code, main_globals, None,
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/runpy.py", line 86, in _run_code
      exec(code, run_globals)
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/__main__.py", line 22, in <module>
      from pip._internal.cli.main import main as _main
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/main.py", line 11, in <module>
      from pip._internal.cli.autocompletion import autocomplete
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/autocompletion.py", line 12, in <module>
      from pip._internal.cli.main_parser import create_main_parser
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/cli/main_parser.py", line 9, in <module>
      from pip._internal.build_env import get_runnable_pip
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/build_env/__init__.py", line 8, in <module>
      from pip._internal.build_env.installer import (
    File "/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/build_env/installer.py", line 14, in <module>
      from pip._internal.exceptions import (
  ImportError: cannot import name 'BuildDependencyInstallError' from 'pip._internal.exceptions' (/opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages/pip/_internal/exceptions.py)
  Error: Process completed with exit code 1.
```

### Screenshots / Screen Recordings

N/A

### Occurrences

Twice
1. https://github.com/oppia/oppia/actions/runs/30485837861/job/90715097663?pr=26962
2. https://github.com/oppia/oppia/actions/runs/30485837861/job/90715097732?pr=26962

### Additional Information

N/A

### Debugging document link

_No response_

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -45,7 +45,6 @@ runs:
 
     - name: Show where pip installs its packages, and the contents of that folder.
       run: |
-        pip --version
         pip show certifi
         ls -la /opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages
         ls -la /opt/hostedtoolcache/Python/3.10.16/x64/bin
@@ -82,6 +81,23 @@ runs:
           /home/runner/work/oppia/oppia/node_modules
           /home/runner/work/oppia/oppia/third_party/python_libs
           /home/runner/work/oppia/oppia/third_party/static
+
+    - name: Check Yarn Cache
+      run: |
+        # Move outside repository root to ignore the configs in
+        # oppia/.yarnrc. These configs specify that yarn should use the
+        # Oppia-managed yarn executable, which hasn't been downloaded
+        # yet.
+        cd ..
+        # Create a new .yarnrc file to check what is already in the
+        # Oppia-managed yarn cache.
+        echo 'cache-folder "yarn_cache"' > .yarnrc
+        yarn cache list
+        # Delete the .yarnrc file that we just created.
+        rm .yarnrc
+        # Move back to the repository root.
+        cd -
+      shell: bash
 
     - name: Print requirements files (before)
       run: |

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -50,6 +50,23 @@ runs:
         ls -la /opt/hostedtoolcache/Python/3.10.16/x64/bin
       shell: bash
 
+    - name: Check Yarn Cache
+      run: |
+        # Move outside repository root to ignore the configs in
+        # oppia/.yarnrc. These configs specify that yarn should use the
+        # Oppia-managed yarn executable, which hasn't been downloaded
+        # yet.
+        cd ..
+        # Create a new .yarnrc file to check what is already in the
+        # Oppia-managed yarn cache.
+        echo 'cache-folder "yarn_cache"' > .yarnrc
+        yarn cache list
+        # Delete the .yarnrc file that we just created.
+        rm .yarnrc
+        # Move back to the repository root.
+        cd -
+      shell: bash
+
     # Production pip dependencies are installed in a custom folder rather than
     # the pip cache (see "Caching Python dependencies" in the README at
     # https://github.com/actions/setup-python?tab=readme-ov-file).
@@ -81,23 +98,6 @@ runs:
           /home/runner/work/oppia/oppia/node_modules
           /home/runner/work/oppia/oppia/third_party/python_libs
           /home/runner/work/oppia/oppia/third_party/static
-
-    - name: Check Yarn Cache
-      run: |
-        # Move outside repository root to ignore the configs in
-        # oppia/.yarnrc. These configs specify that yarn should use the
-        # Oppia-managed yarn executable, which hasn't been downloaded
-        # yet.
-        cd ..
-        # Create a new .yarnrc file to check what is already in the
-        # Oppia-managed yarn cache.
-        echo 'cache-folder "yarn_cache"' > .yarnrc
-        yarn cache list
-        # Delete the .yarnrc file that we just created.
-        rm .yarnrc
-        # Move back to the repository root.
-        cd -
-      shell: bash
 
     - name: Print requirements files (before)
       run: |

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -39,6 +39,8 @@ runs:
         retry_on: error
         shell: bash
         timeout_minutes: 30
+        # Note that the effects of this don't seem to be cached by setup-python,
+        # resulting in the occasional "No module named PIL" errors in CI.
         command: python -m scripts.install_python_dev_dependencies
 
     - name: Show where pip installs its packages, and the contents of that folder.

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -32,27 +32,13 @@ runs:
         echo "${{ steps.setup_python.outputs.python-version }}" > /home/runner/work/.pyversion
       shell: bash
 
-    - name: Cache Python dev dependencies
-      id: restore_python_dev_dependencies_cache
-      uses: actions/cache@v4
-      # We need to include install_python_dev_dependencies.py in the cache key
-      # because it has the version numbers for pip, pip-tools and setuptools.
-      with:
-        key: ${{ runner.os }}-devdeps-python-${{ hashFiles('/home/runner/work/.pyversion', 'requirements_dev.txt', 'scripts/install_python_dev_dependencies.py') }}
-        path: |
-          /opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages
-          /opt/hostedtoolcache/Python/3.10.16/x64/bin
-
     - name: Install third-party Python dev dependencies
-      if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit != 'true'
       uses: nick-fields/retry@v3
       with:
         max_attempts: 3
         retry_on: error
         shell: bash
         timeout_minutes: 30
-        # Note that the effects of this don't seem to be cached by setup-python,
-        # resulting in the occasional "No module named PIL" errors in CI.
         command: python -m scripts.install_python_dev_dependencies
 
     - name: Show where pip installs its packages, and the contents of that folder.

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -38,33 +38,13 @@ runs:
       # We need to include install_python_dev_dependencies.py in the cache key
       # because it has the version numbers for pip, pip-tools and setuptools.
       with:
-        key: ${{ runner.os }}-proddeps-${{ hashFiles('/home/runner/work/.pyversion', 'requirements_dev.txt', 'scripts/install_python_dev_dependencies.py') }}
+        key: ${{ runner.os }}-devdeps-python-${{ hashFiles('/home/runner/work/.pyversion', 'requirements_dev.txt', 'scripts/install_python_dev_dependencies.py') }}
         path: |
-          /opt/hostedtoolcache/Python/3.10.16/x64
-
-    # TODO (#22567): Fix pip not working properly when fetched from cache.
-    - name: Check pip installation and fix if needed
-      run: |
-        set -e
-        echo "Checking if pip is properly installed..."
-
-        # This shows pip-related environment and configuration info for debugging:
-        # https://pip.pypa.io/en/stable/cli/pip_debug/
-        if ! python -m pip debug; then
-          echo "pip is broken. Clearing pip cache..."
-          python -m pip cache purge || true
-          echo "Reinstalling pip and dev dependencies via install_python_dev_dependencies.py..."
-          curl -sS https://bootstrap.pypa.io/get-pip.py -o /tmp/get-pip.py
-          python -m pip uninstall pip -y
-          python /tmp/get-pip.py pip==25.3
-          python -m scripts.install_python_dev_dependencies
-        else
-          echo "pip is working correctly."
-        fi
-      shell: bash
+          /opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages
+          /opt/hostedtoolcache/Python/3.10.16/x64/bin
 
     - name: Install third-party Python dev dependencies
-      if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit == false
+      if: steps.restore_python_dev_dependencies_cache.outputs.cache-hit != 'true'
       uses: nick-fields/retry@v3
       with:
         max_attempts: 3
@@ -82,23 +62,6 @@ runs:
         ls -la /opt/hostedtoolcache/Python/3.10.16/x64/bin
       shell: bash
 
-    - name: Check Yarn Cache
-      run: |
-        # Move outside repository root to ignore the configs in
-        # oppia/.yarnrc. These configs specify that yarn should use the
-        # Oppia-managed yarn executable, which hasn't been downloaded
-        # yet.
-        cd ..
-        # Create a new .yarnrc file to check what is already in the
-        # Oppia-managed yarn cache.
-        echo 'cache-folder "yarn_cache"' > .yarnrc
-        yarn cache list
-        # Delete the .yarnrc file that we just created.
-        rm .yarnrc
-        # Move back to the repository root.
-        cd -
-      shell: bash
-
     # Production pip dependencies are installed in a custom folder rather than
     # the pip cache (see "Caching Python dependencies" in the README at
     # https://github.com/actions/setup-python?tab=readme-ov-file).
@@ -111,11 +74,6 @@ runs:
     # Note that, in the case of a cache miss, the cache is populated when the
     # job completes successfully. See
     # https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/caching-dependencies-to-speed-up-workflows#cache-hits-and-misses
-    - name: Write the exact Python version to a file for cache-busting.
-      run: |
-        echo "${{ steps.setup_python.outputs.python-version }}" > /home/runner/work/.pyversion
-      shell: bash
-
     - name: Cache third_party and node_modules
       id: restore_third_party_and_node_modules_cache
       uses: actions/cache@v4
@@ -145,7 +103,7 @@ runs:
       shell: bash
 
     - name: Install third-party libs on cache miss (retries on error)
-      if: steps.restore_third_party_and_node_modules_cache.outputs.cache-hit == false
+      if: steps.restore_third_party_and_node_modules_cache.outputs.cache-hit != 'true'
       uses: nick-fields/retry@v3
       with:
         max_attempts: 3
@@ -163,7 +121,6 @@ runs:
       shell: bash
 
     - name: Check Yarn Cache
-      # Check the yarn cache again to see what packages were just
-      # installed.
+      # Check the yarn cache to see what packages were installed.
       run: yarn cache list
       shell: bash

--- a/.github/actions/merge-develop-and-set-up-dependencies/action.yml
+++ b/.github/actions/merge-develop-and-set-up-dependencies/action.yml
@@ -57,6 +57,7 @@ runs:
 
     - name: Show where pip installs its packages, and the contents of that folder.
       run: |
+        pip --version
         pip show certifi
         ls -la /opt/hostedtoolcache/Python/3.10.16/x64/lib/python3.10/site-packages
         ls -la /opt/hostedtoolcache/Python/3.10.16/x64/bin

